### ISSUE TITLE
bug:  Fast button component not added to tab queue in fast dialog by default

### ISSUE THREAD

**Issue by scomea** [microsoft/fast#4035]
Fast button elements that are children of a fast dialog don't appear in the dialog's tab queue because fast buttons are not detected as [tabbable](https://github.com/focus-trap/tabbable) because the button tag is not discovered because it exists in the shadow dom.

A possible solution is to mark fast-button elements with a tabindex of "0" by default so tabbable discovers it.  This may affect other web components that hide focusable elements in the shadow DOM and aren't marked wiht a positive tab index.

Example:

1. Create a modal fast dialog with several fast-button children
2. Try to tab between them, note that it doesn't work
3. Add tabindex="0" attributes to the buttons
4. Try tabbing between them, note that it does work

Fast buttons should show up in dialog tab queues by default.

**Comment by scomea** [comment]
Related issue to be aware of - https://github.com/microsoft/fast/pull/4036

**Comment by SethDonohue** [comment]
For now are we sticking with the author needing to add a non-negative tabindex to the fast-buttons for it to show in the dialog tab queue, or are we thinking of changing fast-button to have a tabindex="0" by default? Just curious if there has been any movement on this.

**Comment by chrisdholt** [comment]
> For now are we sticking with the author needing to add a non-negative tabindex to the fast-buttons for it to show in the dialog tab queue, or are we thinking of changing fast-button to have a tabindex="0" by default? Just curious if there has been any movement on this.

I think when @EisenbergEffect lands the Provider DSL we'll be able to plug into that and leverage some of the benefits for our own trap focus implementation. The problem here is that we need to maintain a list of all HTML focusable elements as well as a list of our focusable custom elements, while also providing a method for users to register their own. Add to this the fact that someone may register our custom elements with a namespace other than "fast". I think the provider DSL will provide a hook for us to handle all of that fairly elegantly. For now, because that's not landed, we'll keep the code as is using `tabbable`, which requires tabindex="0" for custom elements.

**Comment by m4thieulavoie** [comment]
@chrisdholt are you suggesting people manually add `tabindex="0"` in their own app for now, or that FAST handles that internally (by adding `tabindex="0"` if none was specified by the user)? If the latter, I'm happy to contribute if you want 👍

**Comment by SethDonohue** [comment]
For now, we need to add tabindex="0" manually. The work he mentioned above with the Provider DSL will help with this issue once it is ready.

**Comment by chrisdholt** [comment]
> @chrisdholt are you suggesting people manually add `tabindex="0"` in their own app for now, or that FAST handles that internally (by adding `tabindex="0"` if none was specified by the user)? If the latter, I'm happy to contribute if you want 👍

@m4thieulavoie yeah, for added clarity on what @SethDonohue mentioned, there are a few considerations that make this difficult to handle all up currently. Currently we use tabbable which doesn't have a method for accounting for custom elements. We need to create a utility for trapping focus and while we could likely handle anything with a `fast` prefix today, the problem really starts in `foundation` prior to things being registered. 

The DSL work will ultimately provide an avenue for us to account for any registered instance of "button" for example. Without that, we can account for elements such as `fast-button`, but what happens if I register as `fluent-button` or `baz-button` - we need to account for the larger story there. The other bit that we need to account for is custom elements which are focusable that aren't part of our package; the registration will be helpful here as well since we can provide a method for folks to opt in.

I'll see if I can work up a task tracking the creation of the trap focus utility with requirements and link that here.

Currently tabbable provides a (less than ideal for us) method of bringing focus. Hopefully the Provider updates will be in soon and we can get rolling on a more elegant update here!

**Comment by scomea** [comment]
How about we do this to lessen some of the pain for dialog:

- add attributes/properties to dialog for the start/end elements of the tab queue that authors can set, either by passing an element reference or id.
- use the provided elements to manage the queue, but keep the tabbable system in place so we don't break existing implementations.

Here's a working branch:
https://github.com/microsoft/fast/compare/users/scomea/dialog-focus-trap

**Comment by radium-v** [comment]
The latest version of Tabbable also includes `focusable` and `isFocusable` functions that I'm using in the toolbar component. For some reason if you wait for the element's `$fastController.isConnected`, then `isFocusable` seems to work with components.

```ts
protected slottedToolbarItemsChanged(prev: unknown, next: HTMLElement[]): void {
    if (this.$fastController.isConnected) {
        this.focusableElements = next.filter(
            async (element: HTMLElement & FASTElement) => {
                if (element.$fastController) {
                    await element.$fastController.isConnected;
                }
                return isFocusable(element);
            }
        );
    }
}
```

**Comment by scomea** [comment]
@radium-v  Even without setting tab index to 0?

**Comment by radium-v** [comment]
@scomea The experiment in my previous comment actually broke tests for proceeding elements in our test suite (the `async` part, though the check for `$fastController` may still be viable).

One possible way I've been thinking about could be to look for a specific `ref` property - so far `control` seems to be the most prevalent in foundation. Other than that, I haven't been able to find any intrinsic way to comfortably assume that an element is tabbable/focusable.

There are some major issues when thinking about how to see if an element's internals are focusable/tabbable:
- Components with `delegatesFocus: true` may not be directly focusable themselves, but may have internals that are focusable
- Components with `mode: closed` can't be searched with `shadowRoot.querySelector()`

---

For reference, the `isFocusable` function in Tabbable [checks these things](https://github.com/focus-trap/tabbable/blob/master/src/index.js#L164):
- one of an explicit list of matching query strings
- a `<details>` with a `<summary>`
- an `<input>` of type `hidden`
- disabled property
- hidden property

The `isTabbable` function additionally checks:
- not a [non-tabbable radio](https://github.com/focus-trap/tabbable/blob/master/README.md#more-details)
- a `tabindex` greater than or equal to 0
- special browser-specific cases (`video/audio` controls and `details` in Chrome)

Tabbable also has an issue to track compatibility with web components here: focus-trap/tabbable#50.

**Comment by EisenbergEffect** [comment]
If we do our own implementation, we could add a registry so that any tag could be added to a known list. We could even add a special decorator in foundation that lets us decorate components so that they get added to the registry, or something along those lines. In this was, we could make it work with our components, establish a pattern for people building on top of FAST, and still have a way for arbitrary elements to be added from other libraries.

Important: I have zero knowledge of tabbable and may not understand the problem here at all. So, ignore me if this doesn't make any sense.

**Comment by chrisdholt** [comment]
> If we do our own implementation, we could add a registry so that any tag could be added to a known list. We could even add a special decorator in foundation that lets us decorate components so that they get added to the registry, or something along those lines. In this was, we could make it work with our components, establish a pattern for people building on top of FAST, and still have a way for arbitrary elements to be added from other libraries.
> 
> Important: I have zero knowledge of tabbable and may not understand the problem here at all. So, ignore me if this doesn't make any sense.

This is intriguing to me because it could scale across multiple components. The registry knows what is focusable. The registry is used for Dialog (traps focus) Flyout (non-modal dialog traps focus), toolbar, OR any other component someone else makes that *we* don't know about.

**Comment by scomea** [comment]
Sorry, I think a registry would be fine if we only had to worry about a narrow list of component tags but I get the impression the global list of web components could get huge and make the approach unwieldly.  And then there's the closed mode problem. The more I look at this the more I think that any good solution is going to require platform help to give us a better hooks into the tab queue (which we should push for). Until that happens my take is that the most reliable approaches may be to require devs to help identify the first/last items in trapped focus queues they implement via element references or provide a function call.

**Comment by scomea** [comment]
What I propose is this.
-	Any base component that implements a contained tab queue, such as dialog, exposes two attributes/props that correspond to the first and last elements of the tab queue.
-	If it gets a string we look for an element id in the light dom.  This will actually cover most cases, be fast and relatively bulletproof.
-	If it gets an element reference we use that.
-	If it gets a function we invoke it with the containing dom element (ie. the dialog) as a parameter.   This is how folks can hook up calls to other libraries like tabbable.
-	We set it up so that component library implementers can define defaults in template creation (ie. fast and fluent could use different default functions).

**Comment by stefcameron** [comment]
👋 Hi folks, I'm the [tabbable](https://github.com/focus-trap/tabbable) maintainer. I noticed our issue https://github.com/focus-trap/tabbable/issues/50 had been tagged by this issue. Interesting problem! Disclaimer: While I do understand the basic concept of web components and the shadow DOM, I don't know FAST, and I have no experience really working with web components.

I like what @scomea is proposing ⬆️ . It feels light-touch, feels like it deals well with the "black box" that web components are supposed to provide (as I understand them so far), seems to deal with closed or open mode, and sounds like it would enable the use of `focusTrap.createFocusTrap(containingDomElement)` (which would ultimately call `tabbable(containingDomElement)`) to create a trap (or just figure out what's tabbable) within your FAST Dialog web component. I may be missing some finer details, but on the surface, it seems like a good plan for the FAST library.

In any case, if there's something that I could help with on the tabbable side (that isn't FAST-specific, but more general to web components and shadow DOM use), then I'm happy to consider it.

**Comment by scomea** [comment]
@stefcameron What if tabbable could be called with a list of additional focusable tags?  
For example  'this.tabbableElements = tabbable(myContainer, ["fast-button", "fast-checkbox"]);
This would help folks with the issue that focusable web components don't get included in your scan.

**Comment by idoros** [comment]
> additional focusable tags?

I don't think it's a problem to implement.

I assume you mean tags that represent elements that have shadow dom that needs to be scanned for potentially tabbable elements, not definitely tabbable elements themselves.

**Comment by scomea** [comment]
@idoros No, this would be additional elements that are focusable.  For example a "<fast-button>" is a focusable element but tabable does not recognize it as such unless we decorate it with a tabindex of 0.  My thinking is that authors could pass in a list of focusable web components included in their project.  The issue of focusable elements within a shadow dom isn't addressed by this (and personally I don't have any good ideas at the moment for that one).

**Comment by stefcameron** [comment]
@scomea How does the browser recognize these elements as being focusable if they aren't natively focusable, and don't have a tabindex?

**Comment by chrisdholt** [comment]
> @scomea How does the browser recognize these elements as being focusable if they aren't natively focusable, and don't have a tabindex?

@stefcameron Certain components such as `fast-button` or `fast-anchor` leverage the delegatesFocus API from the platform to pass focus to elements within the shadowRoot rather than the host itself: https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus

**Comment by scomea** [comment]
Random thought - is scanning for elements that implement "delegatesFocus" a viable approach?

**Comment by EisenbergEffect** [comment]
I don't believe you can get that metadata from an arbitrary web component. It goes into the config but there's no way to get it back out that I know of. We can add APIs to FASTElement to make these discoverable for our own components but I don't think there's a general way to do it. 

My idea behind the registry is that there could be an API to let the developer tell us what elements did this. We could combine that with our own components being self describing and have something that might work automatically for our own components but be extensible for others.

**Comment by idoros** [comment]
Implementating a list of custom elements to add to the candidate tabbable list is easy.

However I would like to implement support for resolving the list of tabbable elements within shadow doms as well. I wonder if a focus delegating element should be included in the result list if we also figure out its internal tabbable shadow elements.

I have no problem implementing an initial support for resolving custom registered elements (if we can agree on the api/implementation with @stefcameron), but would like to think about how future support would work with that api.

**Comment by scomea** [comment]
@idoros Happy to help test/validate any solution you come up with.  I'm not yet clear on how challenging the "scanning shadow doms" issue is, particularly when they are closed.

**Comment by idoros** [comment]
I don't believe there is any automatic solution with closed mode components. The best we can do is imagine some optional interface, like you described above, that such a component can expose to help navigate its internal focus (not sure if that means exposing internal elements references, as that would bypass the "closed mode").

Another issue that I'm thinking about is slotted elements, they are accessible with queries like Tabbable is using today, but their order is wrong as they need to be interleaved into the shadow dom result list.

Anyhow, I will start writing some tests next week to move this forward. and I think we can move the detailed discussion from here...

**Comment by stefcameron** [comment]
> > @scomea How does the browser recognize these elements as being focusable if they aren't natively focusable, and don't have a tabindex?
> 
> @stefcameron Certain components such as `fast-button` or `fast-anchor` leverage the delegatesFocus API from the platform to pass focus to elements within the shadowRoot rather than the host itself: https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus

Thanks for the reference, @chrisdholt . MDN states this is an experimental API, and it's not supported across all browsers tabbable supports. That makes me favor a more explicit solution, like telling tabbable what additional nodes it should consider focusable, rather than trying to infer it, since we couldn't infer it reliably for all supported browsers.

Not all focusable nodes are tabbable, however, so it would probably be more complicated than just giving tabbable a list of additional nodes or queries that return nodes, or a function that returns nodes.

---

__General question:__ Is tabbable, in the context of this Fast library issue, used on its own, or is it the target of discussion because of its implicit use via focus-trap (which uses tabbable to determine tab order within a trap)?

---

> I don't believe there is any automatic solution with closed mode components. The best we can do is imagine some optional interface, like you described above, that such a component can expose to help navigate its internal focus (not sure if that means exposing internal elements references, as that would bypass the "closed mode").

I'm not sure there's much that could be done from the tabbable side for closed mode since that's a black box, on purpose, which inevitably means some type of custom solution, which could end-up exposing internals, so would be best left to a custom solution than for a library like tabbable to be able to discover, defeating the point of "closed mode", as you said.

**Comment by scomea** [comment]
I've pushed up a draft PR with the changes I proposed.  Still need to add tests, but open for comments/feedback.

**Comment by scomea** [comment]
PR is ready for review.

**Comment by chrisdholt** [comment]
I think this has been closed, closing per a previous PR from @scomea

**Comment by chrisdholt** [comment]
> FYI, `tabbable@beta` and `focus-trap@beta` now have shadow DOM support. Please check them out and LMK if there are any issues.

Thanks @stefcameron for following up here! @scomea as FYI on the improvement available from tabbable!
